crypto/BRCryptoKey.c

```diff
@@ -0,0 +1,278 @@
+    //        pkData.count = pkLen
+    //        guard pkData.withUnsafeMutableBytes({ BRKeyPrivKey(&key, $0.baseAddress?.assumingMemoryBound(to: Int8.self), pkLen) }) == pkLen else { return nil }
+    //        key.clean()
+
+    return cryptoKeyCreateInternal(core, CRYPTO_ADDRESS_PARAMS);
+}
+
+extern BRCryptoKey
+cryptoKeyCreateForBIP32BitID (const char *phrase, int index, const char *uri,  const char *words[]) {
+    if (!BRBIP39PhraseIsValid (words, phrase)) return NULL;
+
+
+    UInt512 seed;
+    BRBIP39DeriveKey (seed.u8, phrase, NULL);
+
+    BRKey core;
+    BRBIP32BitIDKey (&core, &seed, sizeof(UInt512), index, uri);
+
+    return cryptoKeyCreateInternal(core, CRYPTO_ADDRESS_PARAMS);
+}
+
+typedef enum {
+    SERIALIZE_PUBLIC_IDENTIFIER = 1,
+    SERIALIZE_PRIVATE_IDENTIFIER = 2
+} BRCryptoKeySerializeIdentifier;
+
+
+extern BRCryptoKey
+cryptoKeyCreateFromSerializationPublic (uint8_t *data, size_t dataCount) {
+    if (dataCount < 1) return NULL;
+    if (SERIALIZE_PUBLIC_IDENTIFIER != (BRCryptoKeySerializeIdentifier) data[0]) return NULL;
+
+    BRKey core;
+
+    return (1 == BRKeySetPubKey (&core, &data[1], dataCount - 1)
+            ? cryptoKeyCreateInternal (core, CRYPTO_ADDRESS_PARAMS)
+            : NULL);
+}
+
+extern size_t
+cryptoKeySerializePublic (BRCryptoKey key, /* ... */ uint8_t *data, size_t dataCount) {
+    size_t publicKeySize = BRKeyPubKey (&key->core, NULL, 0);
+    size_t serializeSize = 1 + publicKeySize;
+
+    if (NULL == data) return serializeSize;
+    if (dataCount < serializeSize) return 0;
+
+    data[0] = SERIALIZE_PUBLIC_IDENTIFIER;
+    BRKeyPubKey (&key->core, &data[1], dataCount);
+    return dataCount;
+}
+
+extern BRCryptoKey
+cryptoKeyCreateFromSerializationPrivate (uint8_t *data, size_t dataCount) {
+    if (dataCount < 1) return NULL;
+    if (SERIALIZE_PRIVATE_IDENTIFIER != (BRCryptoKeySerializeIdentifier) data[0]) return NULL;
+
+    // Get the AddressParms
+    BRAddressParams addressParams = CRYPTO_ADDRESS_PARAMS;
+    BRKey core;
+
+    // Make a string
+    char privKey[dataCount];
+    memcpy (privKey, &data[1], dataCount - 1);
+    privKey[dataCount - 1] = 0;
+
+    if (!BRPrivKeyIsValid (addressParams, privKey)) return NULL;
+
+    return (1 == BRKeySetPrivKey (&core, addressParams, privKey)
+            ? cryptoKeyCreateInternal (core, addressParams)
+            : NULL);
+}
+
+extern size_t
+cryptoKeySerializePrivate (BRCryptoKey key, /* ... */ uint8_t *data, size_t dataCount) {
+    // If we encode the BRAddressParams, we'll need to increate the required dataCount
+    size_t privateKeySize = BRKeyPrivKey (&key->core, NULL, 0, key->coreAddressParams);
+    size_t serializeSize  = 1 + privateKeySize;
+
+    if (NULL == data) return serializeSize;
+    if (dataCount < serializeSize) return 0;
+
+    data[0] = SERIALIZE_PRIVATE_IDENTIFIER;
+
+    // Make a string
+    char privKey[dataCount]; // includes 1 for '\0'
```

Comment:
data is an input, output or both? Trying to understand why we copy it to privKey and then call BRKeyPrivKey